### PR TITLE
WOWL: navbar: add extra menu items behavior

### PR DESCRIPTION
**Pull request by brboi**

### CODE REVIEW

File: addons/wowl/static/src/components/navbar/navbar.ts [hunk 1]
```diff
@@ -16,14 +16,59 @@ export class NavBar extends Component<{}, OdooEnv> {
...
   menuRepo = useService("menus");
 
+  private _debouncedAdapt = utils.debounce(this._adapt.bind(this), 250);
+
   constructor(...args: any[]) {
     super(...args);
     hooks.onMounted(() => {
-      this.env.bus.on("MENUS:APP-CHANGED", this, () => this.render());
+      this.env.bus.on("MENUS:APP-CHANGED", this, () => {
+        this._debouncedAdapt();
```
Comment (ged-odoo) [inline addons/wowl/static/src/components/navbar/navbar.ts]: maybe you should call the non debounced version here, no?

Comment (brboi) [reply addons/wowl/static/src/components/navbar/navbar.ts]: Oh yeah ! I totally should. And also, remove the _render() call here, as _adapt() will call _render() itself.

File: addons/wowl/static/src/components/navbar/navbar.ts [hunk 2]
```diff
@@ -16,14 +16,58 @@ export class NavBar extends Component<{}, OdooEnv> {
...
+      const moreDropdown = this.el!.querySelector<HTMLElement>(".o_menu_sections_more");
+      let width = moreDropdown!.offsetWidth;
+      for (const el of sectionsElements) {
+        if (sectionsAvailableWidth >= width + el.offsetWidth) {
+          width += el.offsetWidth;
+        } else {
+          const sectionId = el.querySelector("[data-section]")!.getAttribute("data-section")!;
+          this._overflowingAppSections.push(sectionId);
+        }
+      }
```
Comment (brboi) [inline addons/wowl/static/src/components/navbar/navbar.ts]: Note: this loop could lead to an unconsistent ordering of the sections items.

image

Comment (brboi) [reply addons/wowl/static/src/components/navbar/navbar.ts]:
```suggestion
      const sections = [...sectionsElements];
      for (const section of sections) {
        if (sectionsAvailableWidth < width + section.offsetWidth) {
          const overflowingSections = sections.slice(sections.indexOf(section));
          this._overflowingAppSections = overflowingSections.map(
            (overflowingSection) =>
              overflowingSection.querySelector("[data-section]")!.getAttribute("data-section")!
          );
          break;
        }
        width += section.offsetWidth;
      }
```

File: addons/wowl/static/src/components/navbar/navbar.ts [hunk 3]
```diff
@@ -18,13 +18,69 @@ export class NavBar extends Component<{}, OdooEnv> {
...
+        i = 0,
+        element = moreDropdown!;
+      do {
+        width += element.offsetWidth;
+        element = sections[i++];
+      } while (sectionsAvailableWidth >= width + element.offsetWidth);
+
+      // Last sections are overflowing
+      this.currentAppSectionsExtra = [];
+      do {
```
Comment (ged-odoo) [inline addons/wowl/static/src/components/navbar/navbar.ts]: shouldn't this be a simple for loop? 
```
for (let element of sections) { ...
```

File: addons/wowl/static/src/components/navbar/navbar.ts [hunk 4]
```diff
@@ -18,13 +18,69 @@ export class NavBar extends Component<{}, OdooEnv> {
...
+    const sectionsMenu = this.el!.querySelector<HTMLElement>(".o_menu_sections")!;
+    const sectionsAvailableWidth = sectionsMenu.offsetWidth;
+    const sectionsTotalWidth = sections.reduce((sum, s) => sum + s.offsetWidth, 0);
+
+    if (sectionsAvailableWidth < sectionsTotalWidth) {
+      // Sections are overflowing, show "more" menu
+      moreDropdown!.classList.remove("d-none");
+
+      // Check which are NOT
+      let width = 0,
```
Comment (ged-odoo) [inline addons/wowl/static/src/components/navbar/navbar.ts]: the odoo style is to write `let` each time.
```js
let a = b;
let c = d;
...
```

File: addons/wowl/static/src/components/navbar/navbar.ts [hunk 5]
```diff
@@ -18,13 +18,69 @@ export class NavBar extends Component<{}, OdooEnv> {
...
+        const currentAppSection = this.currentAppSections.find(
+          (s) => s.id.toString() === sectionId
+        );
+        this.currentAppSectionsExtra.push(currentAppSection!);
+        element = sections[i++];
+      } while (element);
+    }
+
+    // ------- Final rendering -------
+    return this.render();
```
Comment (ged-odoo) [inline addons/wowl/static/src/components/navbar/navbar.ts:86]: onlly if something should be hidden in dropdown more

File: addons/wowl/static/src/components/navbar/navbar.ts [hunk 6]
```diff
@@ -18,13 +18,69 @@ export class NavBar extends Component<{}, OdooEnv> {
 
   constructor(...args: any[]) {
     super(...args);
-    hooks.onMounted(() => {
-      this.env.bus.on("MENUS:APP-CHANGED", this, () => this.render());
+    hooks.onMounted(async () => {
+      await this.render();
```
Comment (ged-odoo) [inline addons/wowl/static/src/components/navbar/navbar.ts]: remove

**Review by ged-odoo — APPROVED** [review]

**Pull request merged**